File: GolangSolutions/String/M792_匹配子序列的单词数.go
/*
	@author Hyperspace
	@date 2022/11/17
	@file M792_匹配子序列的单词数.go
*/

package String

import "sort"

func numMatchingSubseq(s string, words []string) (ans int) {
	pos := [26][]int{}
	for i, c := range s {
		pos[c-'a'] = append(pos[c-'a'], i)
	}

	check := func(s string) bool {
		idx := -1
		for _, c := range s {
			li := pos[c-'a']
			r := sort.SearchInts(li, idx+1)
			if r == len(li) {
				return false
			}
			idx = li[r]
		}

		return true
	}

	for _, word := range words {
		if check(word) {
			ans++
		}
	}

	return
}

File: CPPSolutions/LeetCode/String/M792_匹配子序列的单词数.cpp
#include <bits\stdc++.h>

using namespace std;

class Solution {
public:
    int numMatchingSubseq(string s, vector<string> &words) {
        vector<vector<int>> pos(26);
        for (int i = 0; i < s.size(); i++) pos[s[i] - 'a'].emplace_back(i);
        int ans = 0;
        auto check = [&](string &w) {
            int idx = -1;
            for (char &c: w) {
                auto &li = pos[c - 'a'];
                int r = upper_bound(li.begin(), li.end(), idx) - li.begin();
                if (r == li.size()) return false;
                idx = li[r];
            }
            return true;
        };
        for (auto &word: words) ans += check(word);

        return ans;
    }
};